Build documentation and test plans

paulsalcido opened the pull request:
Don't merge this yet, I've only gotten started.  I want to talk to you about a few items before I go about making some actual structural changes to the code.

paulsalcido commented on lib/Thingiverse.pm:
```diff
@@ -60,8 +52,19 @@ our $client_id          = 'c587f0f2ee04adbe719b';
 our $access_token       = 'b053a0798c50a84fbb80e66e51bba9c4';
```
@perlygatekeeper I want to turn these 'our' objects into required attributes in the class and get rid of these hard-coded access tokens.  If you want to test with real access tokens locally, I would recommend that you add a gpg encrypted file locally or to some test are and use it to load up stuff.  Then, around your tests you would skip if the person couldn't decrypt the file.

perlygatekeeper replied:
Please note, I'm working here:
https://github.com/perlygatekeeper/perlythingkeeper/pull/1/files
Don't merge this man.
Paul • 2:26 PM
Paul Salcido
OK, Steve, I think I have my first pull request ready to go.  All tests are now passing and I'm happy with most of the work.  After that I have plans for two more pull requests:  Documentation and Dependency Injection.  These will be important for a SOLID code base.  Regardless, will talk later.
Paul • 4:32 PM

perlygatekeeper replied:
I like most of what you did.  I wish you hadn't called it "Cleanup" :).

So I am a little confused about whether this is ready to merge or not.

perlygatekeeper replied:
Concerning the access token.  It is only a temporary, private access code, according to the following excerpt from http://www.thingiverse.com/developers/getting-started (near the bottom):

Application Types
There are four kinds of applications that can be built with the Thingiverse API:
Web - An application that operates in a web browser, but outside the Thingiverse site.
Iframe - An application that operates within the Thingiverse site by loading inside of an iframe. These apps can use the Thingiverse Javascript API to present dialogs to the user for common functions.
Mobile - A standalone application that will operate in a mobile or tablet environment.
Desktop - A standalone application that will operate on a desktop computer environment.
Web and Iframe apps will need to specify the URL where the app lives and should include a callback URL.

Your app will be private until you publish it and the Thingiverse moderators review and approve it. While your application is private you can have up to 10 people register and play with it.

perlygatekeeper replied:
You know how you could really help?  Fix Dist Zilla so that versions work.

paulsalcido committed:
Flatten out Moose attributes.

paulsalcido committed:
Eliminate method boilerplate doc

paulsalcido committed:
Client can use lazy_build

Also, required => 1 is not good for lazy attributes.

paulsalcido committed:
Minimal Thingiverse base documentation update

paulsalcido committed:
Build steps and testing from ubuntu raw

I decided to build this in a very raw linux install via the following
steps:

* sudo lxc-create -n thingiverse -t ubuntu
* sudo lxc-start -n thingiverse -d
* # Various step up steps here to get the system up to date with git
  and get the ssh connections for github set up, not important
* # Clone repo
* dzil build
** Now here is where the errors started popping, and I resolved them
   one by one and added the steps to the README.md file.

Now I'm able to do a full build on a clean installation,  I also
snapshotted the box before I started building so that I can test
it in the future a little more easily.

paulsalcido committed:
Build up to running tests

I had to eliminate the Perl=5.8 line as that was causing attempts
at installing the library PerlInterp, which we really, really,
don't want happening.

Tests are running but a huge number of failures.

paulsalcido committed:
Use appropriate method naming of private methods

Also, move the use; methods in tests to use_ok.

The tests are really really far from passing at this point.  Tons
of tests and libs aren't there, but that'll be fixed eventually.

One problem that I caught was that when I changed a private method
in the Thingiverse class, it broke a ton of other classes.  This
is because they are all referencing the private method.  I'm going
to have to fix this, probably via dependency injection, in a future
pull request.

paulsalcido committed:
00_GAT test count

paulsalcido committed:
01_User tests passing.

paulsalcido committed:
ArrayRef misspelling.

paulsalcido committed:
Remove images test from 02_Thing

Images::List not currently done.  Want clean tests.

paulsalcido committed:
Weirdness in test output removed

paulsalcido committed:
03_Collection tests passing.

paulsalcido committed:
04_Tag tests passing.

paulsalcido committed:
05_Category tests passing

paulsalcido committed:
06_File tests passing

paulsalcido committed:
07_Comment tests passing

paulsalcido committed:
08_Image tests passing.

paulsalcido committed:
09_Tag images passing.

paulsalcido committed:
Counts for tests missing them.

perlygatekeeper commented:
Wow, you've done a lot of good work.  I've also got all the tests passing on my box so I worked on the tests pretty hard.  I don't know how easy the merge will be at this point.   The counts were missing for tests 10 and 11 because I was toying with not using counts.  I also discovered that both 04 and 09 are testing Tag.pm, so I need to see which is better or whether they should be merged and then reuse 09 for testing something else.

How do you wish to proceed in a way that we won't be stepping on each others toes, because I really appreciate the help.

paulsalcido commented:
Steve, we can make sure that we don't step on each other's toes if you work in a branch and you push it up on a regular basis.  I'll take a look at mine, try to rebase it onto your new work, and then see what I end up with.

paulsalcido commented:
@perlygatekeeper Do you want to go no counts, because I could change them all in one commit and be done with it.

paulsalcido committed:
02_Thing test counts

paulsalcido committed:
TODO tests now running.
